### PR TITLE
feat: thorchain streaming swap progress bar

### LINKED ISSUE #5191: Create thor streaming progress bar
# Overview

Thorchain streaming swaps take an exceedingly long amount of time to complete (hours), and currently our UI shows the trade as complete at the point of time the streaming begins.

We require a progress bar to indicate to users the progress of the streaming swap such that the dont think their trade is stuck or broken in some way.

# Details

Initially use the chakra UI component `Progress`, and then get @reallybeard to oil it up.

```
import {
  Progress,
} from '@chakra-ui/react'
```

Fetch trade progress from thor node:

`curl https://dev-daemon.thorchain.shapeshift.com/lcd/thorchain/swap/streaming/7AE1D0DB5947ACD5C48435E117E746022DB65280917DCFEC660450A423C7E803`

```
{
  "tx_id": "7AE1D0DB5947ACD5C48435E117E746022DB65280917DCFEC660450A423C7E803",
  "interval": 9,
  "quantity": 20,
  "count": 18,
  "last_height": 12382114,
  "trade_target": "0",
  "deposit": "4985000000",
  "in": "4486500000",
  "out": "278787922"
}
```

### PR DESCRIPTION
**Pull request by woodenfurniture**
## Description

This adds a progress bar to thorchain streaming swaps. 

Includes progress bar and ability to convey the concept of partial fulfilment with reasons for each failure - a value add over viewblock and other interfaces.

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

closes #5191

## Risk

Risk of broken trade confirmation UI

## Testing

* Test streaming swaps and ensure progress bar appears and works as expected
* Test non-streaming swaps and ensure progress bar does not appear

### Engineering

☝️

### Operations

☝️

## Screenshots (if applicable)

Before confirmation
image

On confirmation
image

On streaming data loaded
image

On send tx confirmed
image

Progress increasing, matches viewblock
image
image

Streaming complete
image

### CODE REVIEW

**Comment by woodenfurniture** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #5192** <a href="https://app.graphite.dev/github/pr/shapeshift/web/5192" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/shapeshift/web/5192?utm_source=stack-comment).

File: src/components/MultiHopTrade/hooks/useThorStreamingProgress/useThorStreamingProgress.tsx [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+}
+
+type FailedSwap = {
+  reason: string
+  swapIndex: number
+}
+
+export const useThorStreamingProgress = (
+  txHash: string | undefined,
+  isThorTrade: boolean,
```
Comment (0xApotheosis) [inline src/components/MultiHopTrade/hooks/useThorStreamingProgress/useThorStreamingProgress.tsx:42]:
```suggestion
  isThorStreamingSwap: boolean,
```

File: src/components/MultiHopTrade/hooks/useThorStreamingProgress/useThorStreamingProgress.tsx [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+): {
+  progressProps: ProgressProps
+  attemptedSwaps: number
+  totalSwaps: number
```
Comment (0xApotheosis) [inline src/components/MultiHopTrade/hooks/useThorStreamingProgress/useThorStreamingProgress.tsx:46]: Could improve these names a touch: `attemptedSwapCount` and `totalSwapCount`.

**Review by 0xApotheosis — COMMENTED** [review]
Code-wise looking good, will do a functional pass now.

**Review by 0xApotheosis — CHANGES_REQUESTED** [review]
Super cool feature, though I did end up in this state on a small swap:

Screenshot 2023-09-01 at 12 21 08 pm

The stream status never moved on from loading. It might be because the trade was so small it complete in one step?

**Comment by woodenfurniture** [comment]
> Super cool feature, though I did end up in this state on a small swap:
> 
> Screenshot 2023-09-01 at 12 21 08 pm
> The stream status never moved on from loading. It might be because the trade was so small it complete in one step?

This is intended, the transaction succeeds before the streaming completes. Likely you didnt see the progress bar initialise because it takes about 3-5 mins to pick up the tx from the chain. 

I do think we will need to modify the tx status to not say "completed" while it's still streaming, but I hope we can do that in a follow-up because I'm working on changes that will conflict very heavily with that :P

File: src/assets/translations/en/main.json [hunk 3]
```diff
@@ -676,6 +677,9 @@
     "approvingAsset": "Approving %{symbol}",
     "viewTransaction": "View Transaction",
     "unlimited": "Unlimited",
+    "streamStatus": "Stream Status",
+    "fetchingStreamingSwapData": "Loading...",
```
Comment (gomesalexandre) [inline src/assets/translations/en/main.json:681]: Do we want to improve this copy? Thinking something along the lines of "Initiating swap..."  or "Trade started..."

File: src/components/MultiHopTrade/hooks/useThorStreamingProgress/useThorStreamingProgress.tsx [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+  return {
+    progressProps: {
+      min: 0,
+      max: quantity,
+      value: count,
+      hasStripe: true,
+      isAnimated: !isComplete,
+      colorScheme: isComplete ? 'green' : 'blue',
+    },
+    attemptedSwaps: count ?? 0,
```
Comment (gomesalexandre) [inline src/components/MultiHopTrade/hooks/useThorStreamingProgress/useThorStreamingProgress.tsx:126]: Wondering if this is the best vernacular, these were not attempted but actually completed internal swaps?

File: src/components/MultiHopTrade/hooks/useThorStreamingProgress/useThorStreamingProgress.tsx [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+  quantity?: number
+  count?: number
+  last_height?: number
+  trade_target?: string
+  deposit?: string
+  in?: string
+  out?: string
+  failed_swaps?: number[]
+  failed_swap_reasons?: string[]
+}
```
Comment (gomesalexandre) [inline src/components/MultiHopTrade/hooks/useThorStreamingProgress/useThorStreamingProgress.tsx:20]: Can we make this a union with THOR `ThornodeQuoteResponseError` in case anything goes wrong with either our request generation, or on the THOR side? e.g https://paw.pt/h7dosA29

https://github.com/shapeshift/web/blob/15786601ba3c853d711b83ff268b9b4a377ab6ab/src/lib/swapper/swappers/ThorchainSwapper/types.ts#L51

File: src/components/MultiHopTrade/hooks/useThorStreamingProgress/useThorStreamingProgress.tsx [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+
+export const getThorchainStreamingSwap = async (
+  sellTxHash: string,
+): Promise<ThornodeStreamingSwapResponse | undefined> => {
+  const thorTxHash = sellTxHash.replace(/^0x/, '')
+  const { data: streamingSwapData } = await axios.get<ThornodeStreamingSwapResponse>(
```
Comment (gomesalexandre) [inline src/components/MultiHopTrade/hooks/useThorStreamingProgress/useThorStreamingProgress.tsx:26]: And consume the union here

File: src/components/MultiHopTrade/hooks/useThorStreamingProgress/useThorStreamingProgress.tsx [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+    `${getConfig().REACT_APP_THORCHAIN_NODE_URL}/lcd/thorchain/swap/streaming/${thorTxHash}`,
+  )
+
+  if (!streamingSwapData) return
```
Comment (gomesalexandre) [inline src/components/MultiHopTrade/hooks/useThorStreamingProgress/useThorStreamingProgress.tsx:30]: And handle the errored response here
```suggestion
  if (!streamingSwapData || streamingSwapData.error) return
```

**Comment by gomesalexandre** [comment]
> The stream status never moved on from loading. It might be because the trade was so small it complete in one step?

Tangentially, is this monkey patched? Such a small swap should obviously not show up as a streaming swap option

File: src/components/MultiHopTrade/hooks/useThorStreamingProgress/useThorStreamingProgress.tsx [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+import { getConfig } from 'config'
+import { useEffect, useMemo, useRef, useState } from 'react'
+import { usePoll } from 'hooks/usePoll/usePoll'
+const POLL_INTERVAL_MILLISECONDS = 30_000 // 30 seconds
+
+type ThornodeStreamingSwapResponse = {
+  tx_id?: string
+  interval?: number
+  quantity?: number
+  count?: number
```
Comment (gomesalexandre) [inline src/components/MultiHopTrade/hooks/useThorStreamingProgress/useThorStreamingProgress.tsx:12]: These seem undefined for non-streaming swaps, can we perhaps use them as discriminator on whether or not to use the progress bar / "Loading" copy?

Waited a few minutes to ensure the Tx status is properly picked by the THOR node and still got the same response.

image

This might be an edge case, but may be nice to tackle regardless 🙏🏽

**Review by gomesalexandre — APPROVED** [review]
Tested locally, LGTM minus a few naming/types comments and a possible improvement of the flow. Didn't test the *actual* progress bar with a large swap because something something testing funds, will leave this for ops testing.

### Non-streaming swap

image
image

- Such regular swaps correctly have no "Loading" nor progress bar

### Streaming swaps

Tested with the following monkey-patch

```diff
diff --git a/src/lib/swapper/swappers/ThorchainSwapper/getThorTradeQuote/getTradeQuote.ts b/src/lib/swapper/swappers/ThorchainSwapper/getThorTradeQuote/getTradeQuote.ts
index a2e8e57910..708f7d8d21 100644
--- a/src/lib/swapper/swappers/ThorchainSwapper/getThorTradeQuote/getTradeQuote.ts
+++ b/src/lib/swapper/swappers/ThorchainSwapper/getThorTradeQuote/getTradeQuote.ts
@@ -160,9 +160,7 @@ export const getThorTradeQuote = async (
 
   const perRouteValues = [getRouteValues(swapQuote, false)]
 
-  streamingSwapQuote &&
-    swapQuote.expected_amount_out !== streamingSwapQuote.expected_amount_out &&
-    perRouteValues.push(getRouteValues(streamingSwapQuote, true))
+  streamingSwapQuote && perRouteValues.push(getRouteValues(streamingSwapQuote, true))
 
   const getRouteRate = (expectedAmountOutThorBaseUnit: string) => {
     const THOR_PRECISION = 8
```

image
image
image
image

Can confirm the flow stated by @0xApotheosis above, though I believe this to be an inherent issue of testing this with a monkey-patched streaming swaps for small amounts that aren't actually streaming. Added a possible fix for it in https://github.com/shapeshift/web/pull/5192/files#r1312857963

**Comment by 0xApotheosis** [comment]
> > The stream status never moved on from loading. It might be because the trade was so small it complete in one step?
> 
> Tangentially, is this monkey patched? Such a small swap should obviously not show up as a streaming swap option

Yep, she was indeed monkey patched!

**Review by 0xApotheosis — APPROVED** [review]
Approved as the current issues are intended. Address the suggestions at your discretion.

File: src/assets/translations/en/main.json (hunk 3 above)
Comment (woodenfurniture) [reply src/assets/translations/en/main.json:681]: I opted to remove the copy because its both "loading streaming swap progress" and "unable to load streaming swap metadata". This and the other issues with contradicting statuses will be address in a follow up when we upgrade the thorchain swapper to provide accurate trade status

File: src/components/MultiHopTrade/hooks/useThorStreamingProgress/useThorStreamingProgress.tsx (hunk 5 above)
Comment (woodenfurniture) [reply src/components/MultiHopTrade/hooks/useThorStreamingProgress/useThorStreamingProgress.tsx:20]: 🧠 good suggestion, will do!

File: src/components/MultiHopTrade/hooks/useThorStreamingProgress/useThorStreamingProgress.tsx (hunk 8 above)
Comment (woodenfurniture) [reply src/components/MultiHopTrade/hooks/useThorStreamingProgress/useThorStreamingProgress.tsx:12]: In this case I erred on the side of safety and used the schema from the swagger docs to define them as optional or required, and it seems there are no required fields defined for this response.

https://thornode.ninerealms.com/thorchain/doc/
image

**Comment by woodenfurniture** [comment]
> Tested locally, LGTM minus a few naming/types comments and a possible improvement of the flow. Didn't test the _actual_ progress bar with a large swap because something something testing funds, will leave this for ops testing.
> 
> ### Non-streaming swap
> image image
> * Such regular swaps correctly have no "Loading" nor progress bar
> 
> ### Streaming swaps
> Tested with the following monkey-patch
> 
> ```diff
> diff --git a/src/lib/swapper/swappers/ThorchainSwapper/getThorTradeQuote/getTradeQuote.ts b/src/lib/swapper/swappers/ThorchainSwapper/getThorTradeQuote/getTradeQuote.ts
> index a2e8e57910..708f7d8d21 100644
> --- a/src/lib/swapper/swappers/ThorchainSwapper/getThorTradeQuote/getTradeQuote.ts
> +++ b/src/lib/swapper/swappers/ThorchainSwapper/getThorTradeQuote/getTradeQuote.ts
> @@ -160,9 +160,7 @@ export const getThorTradeQuote = async (
>  
>    const perRouteValues = [getRouteValues(swapQuote, false)]
>  
> -  streamingSwapQuote &&
> -    swapQuote.expected_amount_out !== streamingSwapQuote.expected_amount_out &&
> -    perRouteValues.push(getRouteValues(streamingSwapQuote, true))
> +  streamingSwapQuote && perRouteValues.push(getRouteValues(streamingSwapQuote, true))
>  
>    const getRouteRate = (expectedAmountOutThorBaseUnit: string) => {
>      const THOR_PRECISION = 8
> ```
> 
> image image image image
> Can confirm the flow stated by @0xApotheosis above, though I believe this to be an inherent issue of testing this with a monkey-patched streaming swaps for small amounts that aren't actually streaming. Added a possible fix for it in [#5192 (files)](https://github.com/shapeshift/web/pull/5192/files#r1312857963)

For reference some of the quirks are noted here for fix in follow-up 🙂
https://github.com/shapeshift/web/pull/5192#issuecomment-1702055620
